public/app/features/trails/ActionTabs/BreakdownScene.tsx

```diff
@@ -377,12 +381,14 @@ function buildNormalLayout(queryDef: AutoQueryDef) {
       maxDataPoints: 300,
       queries: queryDef.queries,
     }),
-    options: [
+    breakdownLayoutOptions: [
       { value: 'single', label: 'Single' },
       { value: 'grid', label: 'Grid' },
       { value: 'rows', label: 'Rows' },
     ],
-    layouts: [
+    activeBreakdownLayout: 'grid',
+    onBreakdownLayoutChange,
```

Comment:
I am not sure if I should pass this function or create an event in LayoutSwitcher and publish it. 
Any suggestions? @darrenjaneczek @matyax ?

Reply:
<details>

<summary>.git patch if you want to try</summary>

```
Subject: [PATCH] fix unit tests
---
Index: public/app/features/trails/ActionTabs/BreakdownScene.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/public/app/features/trails/ActionTabs/BreakdownScene.tsx b/public/app/features/trails/ActionTabs/BreakdownScene.tsx
--- a/public/app/features/trails/ActionTabs/BreakdownScene.tsx	(revision 788561f2a004ca9bdb689692735655dfdc79f034)
+++ b/public/app/features/trails/ActionTabs/BreakdownScene.tsx	(date 1721923704735)
@@ -1,5 +1,5 @@
 import { css } from '@emotion/css';
-import { min, max, isNumber, throttle } from 'lodash';
+import { isNumber, max, min, throttle } from 'lodash';
 
 import { DataFrame, FieldType, GrafanaTheme2, PanelData, SelectableValue } from '@grafana/data';
 import {
@@ -35,9 +35,8 @@
 
 import { AddToFiltersGraphAction } from './AddToFiltersGraphAction';
 import { ByFrameRepeater } from './ByFrameRepeater';
-import { LayoutSwitcher } from './LayoutSwitcher';
+import { BreakdownLayoutChange, LayoutSwitcher } from './LayoutSwitcher';
 import { breakdownPanelOptions } from './panelConfigs';
-import { BreakdownLayoutChangeCallback, BreakdownLayoutType } from './types';
 import { getLabelOptions } from './utils';
 import { BreakdownAxisChangeEvent, yAxisSyncBehavior } from './yAxisSyncBehavior';
 
@@ -93,16 +92,17 @@
     // in the same axis will result in no updates to the panels.
 
     const trail = getTrailFor(this);
-    trail.state.$timeRange?.subscribeToState(() => {
-      // The change in time range will cause a refresh of panel values.
-      this.clearBreakdownPanelAxisValues();
-    });
+    // The change in time range will cause a refresh of panel values.
+    trail.state.$timeRange?.subscribeToState(() => this.clearBreakdownPanelAxisValues);
+
+    this._subs.add(this.subscribeToEvent(BreakdownLayoutChange, () => this.clearBreakdownPanelAxisValues()));
 
     this.updateBody(variable);
   }
 
   private breakdownPanelMaxValue: number | undefined;
   private breakdownPanelMinValue: number | undefined;
+
   public reportBreakdownPanelData(data: PanelData | undefined) {
     if (!data) {
       return;
@@ -180,8 +180,8 @@
 
     if (!variable.state.loading && variable.state.options.length) {
       stateUpdate.body = variable.hasAllValue()
-        ? buildAllLayout(options, this._query!, this.onBreakdownLayoutChange)
-        : buildNormalLayout(this._query!, this.onBreakdownLayoutChange);
+        ? buildAllLayout(options, this._query!)
+        : buildNormalLayout(this._query!);
     } else if (!variable.state.loading) {
       stateUpdate.body = undefined;
       stateUpdate.blockingMessage = 'Unable to retrieve label options for currently selected metric.';
@@ -192,10 +192,6 @@
     this.setState(stateUpdate);
   }
 
-  public onBreakdownLayoutChange = (_: BreakdownLayoutType) => {
-    this.clearBreakdownPanelAxisValues();
-  };
-
   public onChange = (value?: string) => {
     if (!value) {
       return;
@@ -269,11 +265,7 @@
   };
 }
 
-export function buildAllLayout(
-  options: Array<SelectableValue<string>>,
-  queryDef: AutoQueryDef,
-  onBreakdownLayoutChange: BreakdownLayoutChangeCallback
-) {
+export function buildAllLayout(options: Array<SelectableValue<string>>, queryDef: AutoQueryDef) {
   const children: SceneFlexItemLike[] = [];
 
   for (const option of options) {
@@ -325,7 +317,6 @@
       { value: 'rows', label: 'Rows' },
     ],
     activeBreakdownLayout: 'grid',
-    onBreakdownLayoutChange,
     breakdownLayouts: [
       new SceneCSSGridLayout({
         templateColumns: GRID_TEMPLATE_COLUMNS,
@@ -346,7 +337,7 @@
 
 const GRID_TEMPLATE_COLUMNS = 'repeat(auto-fit, minmax(400px, 1fr))';
 
-function buildNormalLayout(queryDef: AutoQueryDef, onBreakdownLayoutChange: BreakdownLayoutChangeCallback) {
+function buildNormalLayout(queryDef: AutoQueryDef) {
   const unit = queryDef.unit;
 
   function getLayoutChild(data: PanelData, frame: DataFrame, frameIndex: number): SceneFlexItem {
@@ -387,7 +378,6 @@
       { value: 'rows', label: 'Rows' },
     ],
     activeBreakdownLayout: 'grid',
-    onBreakdownLayoutChange,
     breakdownLayouts: [
       new SceneFlexLayout({
         direction: 'column',
@@ -436,6 +426,7 @@
 interface SelectLabelActionState extends SceneObjectState {
   labelName: string;
 }
+
 export class SelectLabelAction extends SceneObjectBase<SelectLabelActionState> {
   public onClick = () => {
     const label = this.state.labelName;
Index: public/app/features/trails/ActionTabs/LayoutSwitcher.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/public/app/features/trails/ActionTabs/LayoutSwitcher.tsx b/public/app/features/trails/ActionTabs/LayoutSwitcher.tsx
--- a/public/app/features/trails/ActionTabs/LayoutSwitcher.tsx	(revision 788561f2a004ca9bdb689692735655dfdc79f034)
+++ b/public/app/features/trails/ActionTabs/LayoutSwitcher.tsx	(date 1721923601914)
@@ -1,4 +1,4 @@
-import { SelectableValue } from '@grafana/data';
+import { BusEventBase, SelectableValue } from '@grafana/data';
 import {
   SceneComponentProps,
   SceneObject,
@@ -13,13 +13,20 @@
 import { reportExploreMetrics } from '../interactions';
 import { MakeOptional, TRAIL_BREAKDOWN_VIEW_KEY } from '../shared';
 
-import { isBreakdownLayoutType, BreakdownLayoutChangeCallback, BreakdownLayoutType } from './types';
+import { BreakdownLayoutType, isBreakdownLayoutType } from './types';
+
+export class BreakdownLayoutChange extends BusEventBase {
+  constructor(public layout: BreakdownLayoutType) {
+    super();
+  }
+
+  public static type = 'breakdown-layout-change';
+}
 
 export interface LayoutSwitcherState extends SceneObjectState {
   activeBreakdownLayout: BreakdownLayoutType;
   breakdownLayouts: SceneObject[];
   breakdownLayoutOptions: Array<SelectableValue<BreakdownLayoutType>>;
-  onBreakdownLayoutChange: BreakdownLayoutChangeCallback;
 }
 
 export class LayoutSwitcher extends SceneObjectBase<LayoutSwitcherState> implements SceneObjectWithUrlSync {
@@ -68,7 +75,7 @@
     reportExploreMetrics('breakdown_layout_changed', { layout: active });
     localStorage.setItem(TRAIL_BREAKDOWN_VIEW_KEY, active);
     this.setState({ activeBreakdownLayout: active });
-    this.state.onBreakdownLayoutChange(active);
+    this.publishEvent(new BreakdownLayoutChange(active), true);
   };
 
   public static Component = ({ model }: SceneComponentProps<LayoutSwitcher>) => {

```
</details>

Reply:
I am thinking that since the change in breakdown layout it affects this one specific thing, a callback is fine.

Reply:
Agree.

Reply:
👍 Thanks for the feedback!